QR-Generator  submission - Felipe Cesar

fel-cesar opened the pull request:
The current project meets all the mandatory and optional requirements, also includes the files used to build the test Nodejs API. Even though I can still think of improvements that can be handy, I mention some of them in the “Roadmap” section of the [README.md](https://github.com/fel-cesar/mobile-test#roadmap). I will update this repository in the remaining days If I'm allowed. [There is a release](https://github.com/fel-cesar/mobile-test/releases/tag/1.0) locking up this initial submission.

fel-cesar committed:
feat: 🎸 Initial Commmit

Basic project generated by flutter and updated '.gitignore' file

fel-cesar committed:
style: 💄 Add lint related files

Added all_lint_rules.yaml and analysis_options.yaml files responsible
for the lint tules applied by the analyser

fel-cesar committed:
feat: 🎸 Add outer initial setup and basic stub screens

fel-cesar committed:
feat: 🎸 Add immutable objects basic structure and lib

fel-cesar committed:
fix: 🐛 Minor bug fix in tests & small code formatting

fel-cesar committed:
feat: 🎸 Add basic 'Clean Code' domain files

fel-cesar committed:
feat: 🎸 Add QrSeedData ValueObject & respective tests

fel-cesar committed:
feat: 🎸 Add QrSeedExpirationDate ValueObject & tests

fel-cesar committed:
feat: 🎸 Add QrSeed entity

fel-cesar committed:
feat: 🎸 Add QrSeedDto & respective tests

fel-cesar committed:
feat: 🎸 Add repository and data sources interface

fel-cesar committed:
feat: 🎸 Add service locator initial structure

fel-cesar committed:
feat: 🎸 Add NetfworkInfo source interface

fel-cesar committed:
feat: 🎸 Add data sources initial implementation

fel-cesar committed:
feat: 🎸 Add implementation of QrSeedRepository with Net. Check

Add the QrSeedRepository implementation alongside with the Network check
and its implementations

fel-cesar committed:
feat: 🎸 Add missing dependencies and respective configuration

fel-cesar committed:
fix: 🐛 Injectable shared preferences not working & others

Injectable shared preferences not working, also missing injectables
annotations in general

fel-cesar committed:
feat: 🎸 Initial Generate QrCode screen

fel-cesar committed:
feat: 🎸 Better bloc state & event management

Changed the bloc event and state structure to better fit the current
requirements. Also added the countdown text widget to the
DisplayQrCodeScreen

fel-cesar committed:
feat: 🎸 Add local network check in display_qr_code_bloc

Added an aditional check for connectivity before trying to get a new
QrCodeSeed when the previous one has been expired. The first request
though, needs an internet connection.

fel-cesar committed:
fix: 🐛 Endpoint in QrCodeRemoteDataSourceImpl

The previous endpoint was returning a stub data that was being sent as a
query parameter. The new endpoint reflects the provided OpenApi
specification, and was built using Nodejs

fel-cesar committed:
feat: 🎸 Add and 'stub' methods to handle QrCode validation

Added structure to handle the validation of the code once its data has
been read. As for now, the response is always true, hence we dont have a
QR code validation policy or endpoint yet. When the validation
requirements are specified the stub should be replaced.

fel-cesar committed:
test: 💍 test_helper.dart file handle mocks, injection & locator

The test helper file has been improved to handle the services mocks,
taking advantage of the locator and dependency injection.

fel-cesar committed:
test: 💍 i_network_info.dart test

fel-cesar committed:
test: 💍 Implement i_local_data source tests

fel-cesar committed:
test: 💍 Implement i_remote_data_source tests

fel-cesar committed:
test: 💍 Implement display_qr_code_bloc tests

fel-cesar committed:
style: 💄 Enforcing linter rules & formatting

In some cases in test files, we have INTENTIONAL unawaited futures. This
could be solved with the pendantic package, although there is a concept
mismatch with the new flutter version.

fel-cesar committed:
style: 💄 Remove unused file

fel-cesar committed:
feat: 🎸 Failure treatment for display_qr_code_screen

fel-cesar committed:
style: 💄 Add superformula colors pallet

Add superformula primary, secondary and background colors to be used
among the remaining widgets of the app. Also added minor UI changes in
the countdown Text widget.

fel-cesar committed:
feat: 🎸 Improve home screen UI/UX and navigation menu

Improve home screen UI/UX to have some content instead of an empty page.
Refactored the navigation menu to be more compliant with the provided
mockups

fel-cesar committed:
feat: 🎸 Scan Qr Code screen basic implementation

fel-cesar committed:
fix: 🐛 Scan qr code incorrect behavior

The scan qr code screen was presenting incorrect behavior. Sometimes it
presented multiple snackbars in sequence, sometimes the valid values
were given as incorrect. The bloc structure and the UI were changed to
fix that.

fel-cesar committed:
docs: ✏️ Update README.md

fel-cesar committed:
style: 💄 Fixing linter and formatting issues

fel-cesar committed:
fix: 🐛 ScanQrCode screen not showing message on camera error

When not being able to access the devices camera, the bloc was not
emmiting any error messages to be displayed by the presentation.

fel-cesar committed:
test: 💍 Add scan_qr_code_bloc tests

fel-cesar committed:
docs: ✏️ Fix grammar and typos in README.md

fel-cesar committed:
chore: 🤖 Add lambda used to create seed endpoint + description

fel-cesar committed:
fix: 🐛 Remove unused stub file

fel-cesar committed:
fix: 🐛 Add permissions for camera access for ios

fel-cesar committed:
style: 💄 Remove commented debug code from lambda

mattsolle commented on README.md:
```diff
@@ -1,32 +1,88 @@
+# Superformula Mobile Test
 
-Make sure you read **all** of this document carefully, and follow the guidelines in it.
+[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-green.svg)](https://conventionalcommits.org) 
```
kudos: We love this approach and use it extensively. Great to see it used here as well.

fel-cesar replied:
Thanks. Since I discovered the **Conventional Commits** approach (about 3 years ago), I instantly fell in love with it. Nice to see that you also use this approach.

mattsolle commented on README.md:
````diff
@@ -1,32 +1,88 @@
 
-## Q&A
+```bash
+$flutter run -d [device_name]
````
chore: this should be changed to `flutter test`.

fel-cesar replied:
That is correct! I will fix it along with all other `chore` comments in this PR.

mattsolle commented on README.md:
```diff
@@ -1,32 +1,88 @@
+- Webpack configuration
+- More related technologies
+
+## Roadmap
+
+ - Integration tests
+ - CI/CD
+ - Scaner Bloc Test
+ - Persistend Seeds from server (actually validate them with another endpoint)
```
kudos: really appreciate calling out where you could go with the project.

jshaw-mgm replied:
chore: some minor spelling errors in the readme.

jshaw-mgm replied:
info: I appreciate the detail found in the readme.  This approach feels very similar to the approach found by [Reso Coder](https://resocoder.com/) whom I am a fan of.

fel-cesar replied:
I always try to write "meaningful documentation",  the README.md ( as the name implies) is the frontline of your repository. It Is the first communication channel with all those interacting with your code. It's nice to see that it was appreciated.

mattsolle commented on lambdas/seed/GET.js:
```diff
@@ -0,0 +1,30 @@
+        characters: ALLOWED_CHARS,
+      }),
+      expires_at: new Date(new Date().getTime() + 1 * 60000).toISOString(),
+      // expire_at: new Date().toISOString(),
```
chore: we can remove the commented out code.

mattsolle commented on lambdas/seed/shared/constants/Errors.js:
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable no-magic-numbers */
+
+const NOT_FOUND = {
+  statusCode: 404,
+  message: 'Resource not found',
+};
+
+const METHOD_UNAVAILABLE = {
+  statusCode: 400,
+  message: 'Unsupported method',
+};
+
+const EXTERNAL_SERVICE = {
+  statusCode: 404,
+  message: 'Resource not found',
+};
```
chore: all three of these are unused and can be removed or utilized inside the `ResponseHelper`'s `createErrorResponse` method.

fel-cesar replied:
Yes, they are used as the `err` parameter in the `ResponseHelper.createErrorResponse()`. Not all of them are being used right, those unused can be removed indeed.

mattsolle commented on lambdas/seed/shared/constants/Errors.js:
```diff
@@ -0,0 +1,28 @@
+  message: 'Unsupported method',
+};
+
+const EXTERNAL_SERVICE = {
+  statusCode: 404,
+  message: 'Resource not found',
+};
```
Suggestion: there are two different handlers for `404`s. One appears to be a 404 response that would be returned and one appears to be a 404 this service would experience internally (when making a call out to a third party). If both were being used, I would suggest separating them.

fel-cesar replied:
You are right, the external service should be removed since we are not making calls to third-party services.

mattsolle commented on lambdas/seed/shared/helpers/ResponseHelper.js:
```diff
@@ -0,0 +1,52 @@
+   */
+  static createErrorResponse(err) {
+    return {
+      statusCode: err.statusCode ? err.statusCode : '400', // default status code
```
bug: we should be passing in a statusCode (or error from the Errors file) instead of defaulting to a `400`. There may be instances we call this function and its an internal error but the user sees a `400` instead of a `500` response.

mattsolle commented on lambdas/seed/shared/helpers/ResponseHelper.js:
```diff
@@ -0,0 +1,52 @@
+   */
+  static createResponse(err, res) {
+    return {
+      statusCode: err ? '400' : '200',
```
bug: we should be passing in a statusCode (or error from the Errors file) instead of defaulting to a `400`. There may be instances we call this function and its an internal error but the user sees a `400` instead of a `500` response.

fel-cesar replied:
That's correct! It will be fixed. The same applies to `createErrorResponse()`

mattsolle commented on lib/application/display_qr_code/display_qr_code_bloc.dart:
```diff
@@ -0,0 +1,53 @@
+  DisplayQrCodeBloc(this._qrSeedRepository, this._networkInfo)
+      : super(const DisplayQrCodeState.initial());
+
+  @override
+  void onChange(Change<DisplayQrCodeState> change) {
+    super.onChange(change);
+    // print(change.toString());
+  }
```
chore: this can be removed since we're not overriding any parent functionality.

mattsolle commented on lib/application/display_qr_code/display_qr_code_bloc.dart:
```diff
@@ -0,0 +1,53 @@
+  @override
+  Future<void> close() {
+    return super.close();
+  }
```
chore: this can be removed since we're not overriding any parent functionality.

mattsolle commented on lib/application/display_qr_code/display_qr_code_bloc.dart:
```diff
@@ -0,0 +1,53 @@
+
+class DisplayQrCodeBloc extends Bloc<DisplayQrCodeEvent, DisplayQrCodeState> {
+  final IQrSeedRepository _qrSeedRepository;
+  final INetworkInfo _networkInfo; // = locator<INetworkInfo>();
```
chore: this commented out code can be removed.

mattsolle commented on lib/application/display_qr_code/display_qr_code_event.dart:
```diff
@@ -0,0 +1,7 @@
+
+@freezed
+class DisplayQrCodeEvent with _$DisplayQrCodeEvent {
+  const factory DisplayQrCodeEvent.requestedNewQrCode() = _RequestedNewQrCode;
+  const factory DisplayQrCodeEvent.qrCodeExpired() = _QrCodeExpired;
```
kudos: Freezed is so clean, love it!

mattsolle commented on lib/application/scan_qr_code/scan_qr_code_bloc.dart:
```diff
@@ -0,0 +1,70 @@
+  void onChange(Change<ScanQrCodeState> change) {
+    super.onChange(change);
+    // print(change.toString());
+  }
```
chore: this can be removed since we're not overriding any parent functionality.

mattsolle commented on lib/application/scan_qr_code/scan_qr_code_bloc.dart:
```diff
@@ -0,0 +1,70 @@
+          final validated = await _qrSeedRepository.validateQrCodeData(ev.code);
+
+          yield* validated.when(left: (left) async* {
+            yield state.copyWith(
+              code: '',
+              isValidating: false,
+              message: 'Server error. Please try again.',
+            );
```
bug: there are multiple `QrSeedFailure` values that can be returned and we should check there value to build out the correct error message. A `connectivityFailure` will return a "Server Error" message when it shouldn't.

fel-cesar replied:
Nice reminder! The bloc and the presentation layer should handle at least the connectivity and the server error differently.

mattsolle commented on lib/domain/core/value_object.dart:
```diff
@@ -0,0 +1,36 @@
+import 'errors.dart';
+import 'failures.dart';
+
+// part 'value_objects.freezed.dart';
```
question: why are we not using Freezed for our ValueObject?

fel-cesar replied:
Usually, I use Freezed for simple Data types and Union types, as seen on bloc states and entities. Since I'm not using all the extra features Freezed provides us in the ValueObject class. I found it easier to use with the @immutable decorator (also simpler to inherit). It can always be changed if necessary. I would be happy to see more elegant solutions, always eager to learn more.

mattsolle commented on lib/domain/display_qr_code/i_qr_seed_local_data_source.dart:
```diff
@@ -0,0 +1,9 @@
+
+abstract class IQrCodeLocalDataSource {
+  /// The key used to store it locally
+  static const cachedQRCode = 'cachedQRCode';
```
chore: `Qr` instead of `QR` to be consistent.

mattsolle commented on lib/domain/display_qr_code/value_objects/qr_seed_expiration_date.dart:
```diff
@@ -0,0 +1,26 @@
+import 'package:superformula_mobile_test/domain/core/value_object.dart';
+import 'package:superformula_mobile_test/domain/core/value_validators.dart';
+
+const seedValidLenght = 32;
```
chore: should be spelled `seedValidLength`

mattsolle commented on lib/infrastructure/display_qr_code/qr_seed_repository.dart:
```diff
@@ -0,0 +1,60 @@
+        return const Either.left(QrSeedFailure.cacheFailure());
+      }
+    }
+  }
```
kudos: super compact and readable.

mattsolle commented on lib/infrastructure/display_qr_code/qr_seed_repository.dart:
```diff
@@ -0,0 +1,60 @@
+class QrSeedRepository implements IQrSeedRepository {
+  @override
+  Future<Either<QrSeedFailure, QrSeed>> getQrCodeSeed() async {
+    final remoteDataSource = locator<IQrCodeRemoteDataSource>();
+    final localDataSource = locator<IQrCodeLocalDataSource>();
+    final networkInfo = locator<INetworkInfo>();
```
chore: I would move the dependencies to the class level (and make them private) to keep consistency with other implementations.

fel-cesar replied:
That actually was left this way intentionally. For the sole purposes of this code challenge, my intentions were to demonstrate that both ways could be done. I prefer to move the dependencies to the class level as well. I will update it with the proper class level dependencies change.

mattsolle commented on lib/infrastructure/display_qr_code/qr_seed_repository.dart:
```diff
@@ -0,0 +1,60 @@
+  @override
+  Future<Either<QrSeedFailure, bool>> validateQrCodeData(String data) async {
+    final remoteDataSource = locator<IQrCodeRemoteDataSource>();
+    final networkInfo = locator<INetworkInfo>();
```
chore: I would move the dependencies to the class level (and make them private) to keep consistency with other implementations.

mattsolle commented on lib/infrastructure/display_qr_code/qrcode_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,41 @@
+    await Future<void>.delayed(const Duration(seconds: 2));
+    // Stub response, always return a valid response
+    // will match every for every data that matches size
+    if (data.length == 32) {
```
chore: we have this length inside of `qr_seed_data` and `qr_seed_expiration_date` as a constant value. We should either point all of these to a single constant or use a constant value here.

fel-cesar replied:
Oh wow, that is something that should be done asap. I really am up for the "no magic numbers" rule, it's amazing how fast those can grow in number when you don't follow that rule. I guess that did not catch my attention due to the fact that was a "stub response". Even though good practices should be applied.Thanks for the reminder

mattsolle commented on lib/infrastructure/display_qr_code/qrcode_local_data_source_impl.dart:
```diff
@@ -0,0 +1,33 @@
+  Future<void> cacheQrSeed(QrSeedDto qrcodeToCache) async {
+    await sharedPreferences.setString(IQrCodeLocalDataSource.cachedQRCode,
+        jsonEncode(qrcodeToCache.toJson()));
+    return;
```
chore: avoid returns in void methods.

mattsolle commented on lib/presentation/core/home_screen.dart:
```diff
@@ -0,0 +1,66 @@
+      backgroundColor: superformulaBackgroundColor,
+      appBar: AppBar(
+        // Here we take the value from the MyHomePage object that was created by
+        // the App.build method, and use it to set our appbar title.
```
chore: comment can be removed since its inaccurate.

mattsolle commented on lib/presentation/core/misc/colors.dart:
```diff
@@ -0,0 +1,29 @@
+      b + ((ds < 0 ? b : (255 - b)) * ds).round(),
+      1,
+    );
+  });
```
suggestion: a comment above this explaining what it is doing would be beneficial.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      // backgroundColor: superformulaBackgroundColor,
```
chore: commented out code can be removed.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+          final bloc = DisplayQrCodeBloc(
+              locator<IQrSeedRepository>(), locator<INetworkInfo>())
+            // ..add(const DisplayQrCodeEvent.started())
+            ..add(const DisplayQrCodeEvent.requestedNewQrCode());
```
chore: other places we're using the locator directly on the Class. I don't know if we can anything by passing in the dependencies on the constructor versus in the Class since we can replace our dependencies with mock ones for testing either way. I would replace this implementation with an internal one to avoid differences in approach.

fel-cesar replied:
That also was left this way intentionally. For the sole purposes of this code challenge, my intentions were to demonstrate that both ways could be done. I will update the code with the proposed fixes.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+        create: (context) {
+          final bloc = DisplayQrCodeBloc(
+              locator<IQrSeedRepository>(), locator<INetworkInfo>())
+            // ..add(const DisplayQrCodeEvent.started())
```
chore: commented out code can be removed.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+            return state.map(
+              initial: (s) => Container(),
+              loadFailure: (s) {
+                // return Text(s.qrcodeFailure.toString());
```
chore: commented out code can be removed.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+                        painter: QrPainter(
+                          data: s.qrcode.seed.getOrCrash(),
+                          version: QrVersions.auto,
+                          // gapless: true,
```
chore: commented out code can be removed.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+                    mainAxisAlignment: MainAxisAlignment.center,
+                    children: [
+                      CustomPaint(
+                        size: const Size.square(size),
```
suggestion: I would just inline the 280. I would expect sizes to be local to where they're used since this is only used one place.

mattsolle commented on lib/presentation/display_qr_code/widgets/expiration_date_timer.dart:
```diff
@@ -0,0 +1,84 @@
+          style: const TextStyle(
+              color: superformulaBackgroundColor,
+              fontWeight: FontWeight.w900,
+              // fontStyle: FontStyle.italic,
```
chore: commented out code can be removed.

mattsolle commented on lib/presentation/display_qr_code/widgets/qr_code_error_widget.dart:
```diff
@@ -0,0 +1,63 @@
+    return Center(
+      child: Stack(
+        alignment: AlignmentDirectional.center,
+        // mainAxisAlignment: MainAxisAlignment.center,
```
chore: commented out code can be removed.

mattsolle commented on lib/presentation/display_qr_code/display_qr_code.dart:
```diff
@@ -0,0 +1,93 @@
+              loadInProgress: (s) {
+                return const Center(child: CircularProgressIndicator());
+              },
+              loadSuccess: (s) {
```
bug: there is an overflow error on landscape for this view.

fel-cesar replied:
Fixed using the OrientationBuilder in (https://github.com/Superformula/mobile-test/pull/53/commits/b3d97e8167d0e02b16c6b9de53c02e1adcf5a65c)

mattsolle commented on test/application/display_qr_code/display_qr_code_bloc_test.dart:
```diff
@@ -0,0 +1,122 @@
+    setUp(registerServices);
+    tearDown(unregisterServices);
+
+    const tSeed = '1';
+    const tExpirationDate = '2020-01-01T00:00:00Z';
+    final tQRCode = QrSeed(
+        seed: QrSeedData(tSeed),
+        expiresAt: QrSeedExpirationDate.withString(tExpirationDate));
+    const tQRCodeFailure = QrSeedFailure.serverFailure();
```
suggestion: I'm not a big fan of Hungarian Notation for variables and [Effective Dart](https://dart.dev/guides/language/effective-dart/style#dont-use-prefix-letters) suggests to not use it.

fel-cesar replied:
The first time I learned mobile development I used Objective-C where is very common to use that kind of notations. Your suggestions is well sustained according to [Effective Dart](https://dart.dev/guides/language/effective-dart/style#dont-use-prefix-letters) and will be reflected in the code.

mattsolle commented on test/application/display_qr_code/display_qr_code_bloc_test.dart:
```diff
@@ -0,0 +1,122 @@
+        await untilCalled(mockRepository.getQrCodeSeed());
+
+        // assert
+        verify(mockRepository.getQrCodeSeed());
```
question: what are your thoughts on using the bloc_test library for this instead?

mattsolle commented on test/application/scan_qr_code/scan_qr_code_bloc_test.dart:
```diff
@@ -0,0 +1,121 @@
+  group('ScanQrCodeBloc -', () {
+    setUp(registerServices);
+    tearDown(unregisterServices);
+    const tSeed = 'd43397d129c3de9e4b6c3974c1c16d1f';
```
chore: same note as above on Hungarian Notation.

mattsolle commented on test/domain/display_qr_code/value_objects_test.dart:
```diff
@@ -0,0 +1,132 @@
+  group('QrSeedData: ', () {
+    setUp(() {
+      // run before all tests
+    });
```
chore: this can be removed.

mattsolle commented on test/infrastructure/display_qr_code/qr_seed_dto_test.dart:
```diff
@@ -0,0 +1,37 @@
+  final tQRCode = QrSeed(
+    seed: QrSeedData(tSeed),
+    expiresAt: QrSeedExpirationDate.withString(tExpirationDate),
+  );
```
chore: same note as above on Hungarian Notation.

mattsolle commented on test/setup/test_helpers.dart:
```diff
@@ -0,0 +1,155 @@
+  return service;
+}
+
+void registerServices() {
+  getAndRegisterIQrSeedRepository();
+  getAndRegisterQRCodeRemoteDataSource();
+  getAndRegisterQRCodeLocalDataSource();
+  getAndRegisterNetworkInfo();
+  getAndRegisterDataConnectionChecker();
+  getAndRegisterSharedPreferences();
+  getAndRegisterClient();
+}
+
+void unregisterServices() {
+  locator.unregister<IQrSeedRepository>();
+  locator.unregister<IQrCodeRemoteDataSource>();
+  locator.unregister<IQrCodeLocalDataSource>();
+  locator.unregister<INetworkInfo>();
+  locator.unregister<DataConnectionChecker>();
+  locator.unregister<SharedPreferences>();
+  locator.unregister<Client>();
+}
```
kudos: beautiful.

mattsolle reviewed (commented):
Hey Felipe!  Thank you for completing the code challenge! It was a pleasure looking over your code. Love the DDD approach! I've spent quiet a bit of time working with Resocoder's approach so your code was incredibly familiar in style. Also, amazing job with the tests! I've left some questions, comments, and bugs for your review.

jshaw-mgm commented on lib/application/display_qr_code/display_qr_code_bloc.dart:
```diff
@@ -0,0 +1,53 @@
+      },
+      qrCodeExpired: (qrcodeExpired) async* {
+        final isConnected = await _networkInfo.isConnected;
+        if (isConnected) {
```
suggestion/bug: if you create a qr code and then disable network connection and wait for it to expire you are able to get past this if statement and see negative numbers for the expiration.


<img width="438" alt="Screen Shot 2021-09-29 at 10 08 06 PM" src="https://user-images.githubusercontent.com/85515499/135373969-e6c44801-d187-4d57-b624-86cf09684216.png">

fel-cesar replied:
Fixed in (https://github.com/Superformula/mobile-test/pull/53/commits/b356b56216404f449d27b545d21355ed6bce5fad)

jshaw-mgm commented on lib/domain/platform/i_network_info.dart:
```diff
@@ -0,0 +1,3 @@
+abstract class INetworkInfo {
+  Future<bool> get isConnected;
+}
```
suggestion: one of the best parts of knowing if the user has a network connection is to inform them that they are experiencing errors because of the lack of connection.  I think it is great this was used proactively to prevent crashes though.  I think informing the user of this would be slightly better.

fel-cesar replied:
Thanks! I improved the Failure handling to inform the user about the different errors. See (https://github.com/Superformula/mobile-test/pull/53/commits/b356b56216404f449d27b545d21355ed6bce5fad) and (https://github.com/Superformula/mobile-test/pull/53/commits/1bb58ede7429ce77a377c6bf25a6aa0cfb5bd60e)

jshaw-mgm commented on README.md:
```diff
@@ -1,32 +1,88 @@
+### Not using the `dartz` package
+Dartz package is an amazing tool with a lot of classes that allow us to better conform to the DDD approach, making error states 'unrepresentable'. Although for this challenge we will e using only a small subset of the entire library. That being said we manually implemented the class `Either` and `ValueObject` to use in this project.
+- Completely opinionated decisions. (avoid using extra libraries whenever possible)
+
```
info: I really appreciate the call out to the fact that some things are opinionated decisions.  It is a subtle yet important part of understanding how to develop software as a part of a team.

jshaw-mgm commented on assets/superformula-logo.svg:
```diff
@@ -0,0 +1,5 @@
+	c-0.1,0-0.3,0.1-0.4,0.2c0,0,0,0,0,0L0.2,13.1c0,0-0.1,0.1-0.1,0.1C0,13.3,0,13.4,0,13.6c0,0.1,0,0.1,0,0.2c0,0.1,0,0.1,0.1,0.2
+	c0,0,0.1,0.1,0.1,0.1l12.4,12.4L0.2,38.8C0,39,0,39.3,0,39.6C0.2,39.8,0.4,40,0.7,40h12.9c0.2,0,0.3-0.1,0.4-0.2l12.7-12.7
+	c0.2-0.1,0.4-0.3,0.4-0.6V13.6C27.2,13.3,27,13.1,26.7,13z M14.2,1.3h10.6L14.2,12V1.3z M12.9,38.7H2.3L12.9,28V38.7z M12.9,24.8
+	L2.3,14.2h10.6V24.8z M12.9,12.9H2.3L12.9,2.3V12.9z M14.2,37.7V27.2h10.5L14.2,37.7z M25.8,25.8H15.3l10.5-10.5V25.8z"></path></svg>
```
info:  nice touch 😅

fel-cesar replied:
Thanks. Always try to deliver the "One more thing...". 😃

jshaw-mgm commented on lib/presentation/display_qr_code/widgets/expiration_date_timer.dart:
```diff
@@ -0,0 +1,84 @@
+  void initState() {
+    _start =
+        widget.expirationDate.getOrCrash().difference(DateTime.now()).inSeconds;
+    startTimer();
```
bug: _start could be initialized to a negative number.

fel-cesar replied:
I've changed (see newer commits) the way we are dealing with the expiration date of QR codes and caching, preventing `_start` from being initialized with a negative value. But there are no countermeasures to avoid it in the `expiration_date_timer.dart` yet. Is that what you meant?

jshaw-mgm commented on README.md:
```diff
@@ -1,32 +1,88 @@
 
-- Good use of structure, security, and performance best practices.
-- Solid testing approach.
-- Extensible code.
```
suggestion:  I don't see any widget tests.  Did I miss them or was this on purpose?

fel-cesar committed:
style: 💄 Unused or commented code, typos, minor style fixes

Fixed typos in README, removed unused code in multiple files, removed
commented out code, fixed 'length' typo, remove unused constants

fel-cesar committed:
style: 💄 Remove Hungarian Case

fel-cesar committed:
fix: 🐛 Remove undefined variables from module export

fel-cesar committed:
fix: 🐛 Fix RequestHelper & ResponseHelper statusCode

fel-cesar committed:
refactor: 💡 Apply standard/standard coding style & minor improv

Better error handling, applied a better coding style, removed unused or
unnecessary code.

fel-cesar committed:
fix: 🐛 Handle different errors during QrCode validation

For different failures that can happen when trying to validate the
scanned QR code, different messages will be emmited by the bloc.

fel-cesar committed:
refactor: 💡 Move dependencies to class level (also private)

fel-cesar committed:
refactor: 💡 Moved bloc dependencies to class level (private)

fel-cesar committed:
fix: 🐛 Layout overflow in display_qr_code landscape mode

Wrapped around an OrientationBuilder and made the QrCode Size
proportional to screen height.

fel-cesar committed:
fix: 🐛 Negative timers and cached qr codes

This fixes the issue where negative timer is displayed when the device
has no connection. The general bloc loginc for handling cached qrcode
has changed, as well as its actions when the current qr code expires.
Different errors are being handled with different messages in the
presentation layer.

fel-cesar committed:
docs: ✏️ Fix typos and add table of contents

fel-cesar committed:
refactor: 💡 (colors) Change loop approach, add documentation

fel-cesar committed:
chore: 🤖 Remove comments and unused imports

fel-cesar committed:
test: 💍 Added widget test for basic widgets
